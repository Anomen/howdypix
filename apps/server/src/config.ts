// Load env variables
require("dotenv").config();

const config = {
  rabbitMq: {
    url: process.env.RABBITMQ_URL || "amqp://localhost"
  },
  serverApi: {
    baseUrl: "",
    port: parseInt(process.env.API_PORT as string) || 3004
  },
  auth: {
    code: {
      secret: "secret_code_string",
      expiry: "2h"
    },
    token: {
      secret: "secret_token_string",
      expiry: "4h"
    },
    refreshToken: {
      secret: "secret_refresh_string",
      expiry: "30d"
    }
  }
};

config.serverApi.baseUrl =
  process.env.API_BASE_URL || `http://localhost:${config.serverApi.port}`;

export default config;
