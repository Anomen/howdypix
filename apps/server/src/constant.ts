export enum EventNames {
  NEW_FILE = "NEW_FILE"
}
