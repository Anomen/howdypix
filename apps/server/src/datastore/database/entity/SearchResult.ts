import {
  Column,
  Entity,
  PrimaryGeneratedColumn,
  ManyToOne,
  CreateDateColumn,
  UpdateDateColumn,
} from "typeorm";
import { Search } from "./Search";
import { Photo } from "./Photo";

@Entity()
export class SearchResult {
  @PrimaryGeneratedColumn()
  id: number;

  @CreateDateColumn()
  createdAt: Date;

  @UpdateDateColumn()
  updatedAt: Date;

  @Column("int")
  searchId!: number;

  @Column("int")
  photoId!: number;

  @Column("int")
  public order!: number;

  @ManyToOne(() => Search, (search) => search.searchResults)
  public search!: Search;

  @ManyToOne(() => Photo, (photo) => photo.searchResults)
  public photo!: Photo;
}
