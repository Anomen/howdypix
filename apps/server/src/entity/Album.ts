import { Column, ViewEntity } from "typeorm";

@ViewEntity({
  expression: `
SELECT
   P.DIR,
   P.SOURCE,
   P.PARENTDIR,
   COUNT(DISTINCT P2.ID) AS NBPHOTOS,
   COUNT(DISTINCT P3.DIR) AS NBALBUMS,
   (
      SELECT
         FILE 
      FROM
         PHOTO P4 
      WHERE
         P4.DIR = P.DIR 
         AND P4.SOURCE = P.SOURCE LIMIT 0,
         1
   )
   AS PREVIEW 
FROM
   PHOTO P 
   LEFT JOIN
      PHOTO P2 
      ON (P2.DIR = P.DIR 
      AND P2.SOURCE = P.SOURCE) 
   LEFT JOIN
      PHOTO P3 
      ON (P3.PARENTDIR = P.DIR 
      AND P3.SOURCE = P.SOURCE) 
GROUP BY
   P.DIR,
   P.SOURCE,
   P.PARENTDIR;
    `,
})
export class Album {
  @Column("text")
  dir: string;

  @Column("text")
  source: string;

  @Column("text")
  parentDir: string;

  @Column("text")
  preview: string;

  @Column("int")
  nbPhotos: number;

  @Column("int")
  nbAlbums: number;
}
