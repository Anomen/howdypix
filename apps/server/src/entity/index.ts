export * from "./Source";
export * from "./Album";
export * from "./Photo";
export * from "./Search";
export * from "./SearchResult";
