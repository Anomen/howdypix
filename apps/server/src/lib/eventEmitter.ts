import StrictEventEmitter from "strict-event-emitter-types";
import { EventEmitter } from "events";
import { ProcessData } from "@howdypix/shared-types";

export interface EventTypes {
  newFile: { root: string; path: string; sourceId: string };
  removeFile: { root: string; path: string; sourceId: string };
  processFile: { root: string; path: string; sourceId: string };
  processedFile: ProcessData;
}

export type Events = StrictEventEmitter<EventEmitter, EventTypes>;
