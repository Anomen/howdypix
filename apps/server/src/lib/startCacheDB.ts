import { Connection } from "typeorm";
import { appDebug, generateThumbnailPaths, hjoin } from "@howdypix/utils";
import { existsSync, statSync, unlinkSync } from "fs";
import { join, parse } from "path";
import { UserConfig } from "../config";
import { Photo } from "../entity/Photo";
import { Events, EventTypes } from "./eventEmitter";
import { Source } from "../entity/Source";
import { Album } from "../entity/Album";

export async function onNewDir(
  { root, hfile }: EventTypes["newDirectory"],
  event: Events,
  connection: Connection
): Promise<void> {
  const absolutePath = join(root, hfile.dir ?? "", hfile.file ?? "");
  const stat = statSync(absolutePath);

  // Check if it exists in the database
  const albumRepository = connection.getRepository(Album);
  const sourceRepository = connection.getRepository(Source);
  const album = await albumRepository.findOne({ where: { inode: stat.ino } });

  // Check the updated_date
  if (!album) {
    const newAlbum = new Album();
    newAlbum.dir = hfile.file ?? "";
    newAlbum.parentDir = hfile.dir ?? "";
    newAlbum.source = hfile.source;
    newAlbum.inode = stat.ino.toString();

    newAlbum.sourceLk =
      (await sourceRepository.findOne({
        source: hfile.source,
      })) ?? null;

    albumRepository.save(newAlbum);
  }
}

export async function onUnlinkDir(
  { root, hfile }: EventTypes["unlinkDirectory"],
  event: Events,
  connection: Connection
): Promise<void> {
  const absolutePath = join(root, hfile.dir ?? "", hfile.file ?? "");
  const stat = statSync(absolutePath);

  // Check if it exists in the database
  const albumRepository = connection.getRepository(Album);
  await albumRepository.delete({ inode: stat.ino.toString() });
}

export async function onNewFile(
  { root, hfile }: EventTypes["newFile"],
  event: Events,
  connection: Connection,
  userConfig: UserConfig
): Promise<void> {
  const absolutePath = join(root, hfile.dir ?? "", hfile.file ?? "");
  const stat = statSync(absolutePath);

  // Check if it exists in the database
  const photoRepository = connection.getRepository(Photo);
  const photo = await photoRepository.findOne({ inode: stat.ino });

  // Check the updated_date
  if (photo?.mtime === stat.mtimeMs) {
    const thumbnailsExist: boolean = generateThumbnailPaths(
      userConfig.thumbnailsDir,
      hfile
    ).reduce(
      (accumulator: boolean, tn) => accumulator && existsSync(tn.path),
      true
    );

    if (thumbnailsExist) {
      return;
    }
  }

  // Act
  event.emit("processFile", { root, hfile });
}

export async function onRemoveFile(
  { hfile }: EventTypes["newFile"],
  event: Events,
  connection: Connection,
  userConfig: UserConfig
): Promise<void> {
  const photoRepository = connection.getRepository(Photo);

  // Remove from database
  await photoRepository.delete(hfile);

  // Remove thumbnails
  generateThumbnailPaths(userConfig.thumbnailsDir, hfile).forEach((tn) => {
    unlinkSync(tn.path);
  });
}

export async function onProcessedFile(
  file: EventTypes["processedFile"],
  event: Events,
  connection: Connection
): Promise<void> {
  const albumRepository = connection.getRepository(Album);
  const photoRepository = connection.getRepository(Photo);

  const photo = new Photo();
  photo.make = file.exif.make ?? "";
  photo.model = file.exif.model ?? "";
  photo.ISO = file.exif.ISO ?? 0;
  photo.shutter = file.exif.shutter ?? 0;
  photo.processedShutter = file.exif.shutter
    ? Math.round((1 / file.exif.shutter) * 10) / 10
    : 0;
  photo.aperture = file.exif.aperture ?? 0;
  photo.processedAperture = file.exif.aperture
    ? Math.round(file.exif.aperture * 10) / 10
    : 0;
  photo.createDate = file.exif.createDate ?? 0;
  photo.inode = file.stat.inode;
  photo.mtime = file.stat.mtime;
  photo.ctime = file.stat.ctime;
  photo.birthtime = file.stat.birthtime;
  photo.size = file.stat.size;
  photo.source = file.hfile.source;
  photo.parentDir = parse(file.hfile.dir ?? "").dir ?? "";
  photo.dir = file.hfile.dir ?? "";
  photo.file = file.hfile.file ?? "";

  photo.album =
    (await albumRepository.findOne({
      dir: file.hfile.dir ?? "",
      source: file.hfile.source,
      parentDir: parse(file.hfile.dir ?? "").dir ?? "",
    })) ?? null;

  await photoRepository.save(photo);
  appDebug("db")(`Saved ${hjoin(file.hfile)}.`);
}

export async function startCacheDB(
  event: Events,
  userConfig: UserConfig,
  connection: Connection
): Promise<Connection> {
  // Insert the source entries in the database
  await Promise.all(
    Object.keys(userConfig.photoDirs).map(async (source) => {
      await Source.upsert(source, userConfig.photoDirs[source]);
    })
  );

  event.on("newFile", (params) =>
    onNewFile(params, event, connection, userConfig)
  );

  event.on("removeFile", (params) =>
    onRemoveFile(params, event, connection, userConfig)
  );

  event.on("newDirectory", (params) => onNewDir(params, event, connection));

  event.on("unlinkDirectory", (params) =>
    onUnlinkDir(params, event, connection)
  );

  event.on("processedFile", (file) => onProcessedFile(file, event, connection));

  return connection;
}
