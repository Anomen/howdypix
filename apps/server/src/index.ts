import "reflect-metadata";
import config from "./config";
import { applyApolloMiddleware } from "./middleware/apollo";
import { loadUserConfig } from "./lib/loadUserConfig";
import { startFileScan } from "./lib/startFileScan";
import EventEmitter from "events";
import { Events } from "./lib/eventEmitter";
import { startRabbitMq } from "./lib/startRabbitMq";
import { startCacheDB } from "./lib/startCacheDB";
import express from "express";
import { staticHandler } from "./middleware/static";
import { emailListHandler, emailViewHandler } from "./middleware/email";
import { applyAuthMiddleware } from "./middleware/auth";

async function main() {
  const event: Events = new EventEmitter();

  const userConfig = loadUserConfig();
  console.log("User Configuration loaded:");
  console.log(userConfig);

  /**
   * Start the mandatory services
   */
  await startCacheDB(event, userConfig);
  await startRabbitMq(event, userConfig, config.rabbitMq.url);
  await startFileScan(event, userConfig);

  /**
   * Start the API
   */
  const app = express();

  app.use(express.json());

  app.get("/static/*", staticHandler);

  // Only when we develop: it's easier to check the email templates
  // and develop them.
  app.get("/email", emailListHandler);
  app.get("/email/*", emailViewHandler);

  // Authentication routes
  applyAuthMiddleware(app);

  // Attach the Apollo middlewares
  applyApolloMiddleware(app, userConfig);

  app.listen({ port: config.serverApi.port }, () => {
    console.log(`API server started on port ${config.serverApi.port}.`);
  });
}

main();
