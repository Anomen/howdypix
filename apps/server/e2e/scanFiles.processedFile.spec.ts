import waitForExpect from "wait-for-expect";
import { statSync } from "fs";
import { ProcessData } from "@howdypix/shared-types";
import { appError } from "@howdypix/utils";
import { startCacheDB } from "../src/lib/startCacheDB";
import { Events } from "../src/lib/eventEmitter";
import { initialize } from "./scanFiles.setup";
import { Photo } from "../src/entity";

jest.mock("@howdypix/utils", () => ({
  ...jest.requireActual("@howdypix/utils"),
  appError: jest.fn(),
}));

jest.mock("fs", () => ({
  ...jest.requireActual("fs"),
  statSync: jest.fn(),
}));

describe("[ProcessedFile] Scanning files", (): void => {
  let events: Events;

  const dir = "album";
  const file = "PHOTO.jpg";

  const baseProcessedFile: ProcessData = {
    hfile: {
      dir,
      file,
      source: "first",
    },
    root: "/root",
    exif: {
      aperture: 0.1,
      createDate: 1,
      ISO: 2000,
      make: "Make",
      model: "Model",
      shutter: 0.23,
    },
    stat: {
      birthtime: 2,
      ctime: 3,
      mtime: 4,
      size: 5,
      inode: 1234,
    },
    thumbnails: ["thumb1", "thumb2", "thumb3"],
  };

  const {
    pathHashCode,
    baseUserConfig,
    retrieveAlbums,
    retrievePhotos,
    resetScanFilesTests,
    connection,
  } = initialize();

  // ===================================================
  // Reset after each tests
  // ===================================================
  beforeEach(
    async (): Promise<void> => {
      jest.resetAllMocks();
      jest.restoreAllMocks();

      const setup = await resetScanFilesTests();
      events = setup.events;

      (statSync as jest.Mock).mockImplementation((name: string) => ({
        ino: pathHashCode(name),
        mtime: new Date(1),
        ctime: new Date(2),
        birthtime: new Date(3),
        size: 10,
      }));

      await startCacheDB(events, baseUserConfig, connection);

      // 1. Add a photo
      events.emit("newFile", {
        hfile: {
          dir,
          file,
          source: "first",
        },
        root: "/root",
      });

      await waitForExpect(
        async (): Promise<void> => {
          // eslint-disable-next-line jest/no-standalone-expect
          expect(await retrieveAlbums()).toHaveLength(3);
          // eslint-disable-next-line jest/no-standalone-expect
          expect(await retrievePhotos()).toHaveLength(1);
        }
      );
    }
  );

  test("should update the entry", async (): Promise<void> => {
    jest.spyOn(connection.getRepository(Photo), "save");

    events.emit("processedFile", baseProcessedFile);

    await waitForExpect(
      async (): Promise<void> => {
        expect(connection.getRepository(Photo).save).toHaveBeenCalled();
      }
    );

    expect(await retrievePhotos()).toMatchSnapshot();
  });

  test("should NOT update if the photo is not in the DB", async (): Promise<
    void
  > => {
    jest.spyOn(connection.getRepository(Photo), "save");

    events.emit("processedFile", {
      ...baseProcessedFile,
      hfile: { ...baseProcessedFile.hfile, file: "DOES NOT EXIST" },
    });

    await waitForExpect(
      async (): Promise<void> => {
        expect(appError).toHaveBeenCalled();
      }
    );

    expect(connection.getRepository(Photo).save).not.toHaveBeenCalled();
  });

  test("should NOT update if the album is not in the DB", async (): Promise<
    void
  > => {
    jest.spyOn(connection.getRepository(Photo), "save");

    events.emit("processedFile", {
      ...baseProcessedFile,
      hfile: { ...baseProcessedFile.hfile, dir: "DOES NOT EXIST" },
    });

    await waitForExpect(
      async (): Promise<void> => {
        expect(appError).toHaveBeenCalled();
      }
    );

    expect(connection.getRepository(Photo).save).not.toHaveBeenCalled();
  });
});
