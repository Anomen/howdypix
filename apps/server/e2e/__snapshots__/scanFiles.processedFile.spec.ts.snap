// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[ProcessedFile] Scanning files should update the entry 1`] = `
Array [
  Photo {
    "ISO": 2000,
    "album": Album {
      "dir": "album",
      "id": 3,
      "inode": "Zmlyc3RhbGJ1bQ==",
      "parentDir": "",
      "photos": undefined,
      "source": "first",
      "sourceLk": undefined,
    },
    "aperture": 0.1,
    "birthtime": 3,
    "createDate": 1,
    "ctime": 2,
    "dir": "album",
    "file": "PHOTO.jpg",
    "id": 1,
    "inode": "YWxidW1QSE9UTy5qcGc=",
    "make": "Make",
    "model": "Model",
    "mtime": 1,
    "parentDir": "",
    "processedAperture": 0.1,
    "processedShutter": 4.3,
    "searchResults": undefined,
    "shutter": 0.23,
    "size": 10,
    "source": "first",
    "status": "processed",
  },
]
`;
